src/openstack_manager.py

```diff
@@ -2,58 +2,45 @@
 # See LICENSE file for licensing details.
 
 """Module for handling interactions with OpenStack."""
+import json
 import logging
-from pathlib import Path
+import secrets
+from dataclasses import dataclass
 
+# subprocess module is used to call image build script.
+from subprocess import SubprocessError  # nosec
+from typing import Iterable, Optional
+
+import jinja2
 import keystoneauth1.exceptions.http
 import openstack
+import openstack.compute.v2.server
 import openstack.connection
 import openstack.exceptions
-import yaml
+import openstack.image.v2.image
+from openstack.exceptions import OpenStackCloudException
 from openstack.identity.v3.project import Project
 
-from errors import OpenStackInvalidConfigError, OpenStackUnauthorizedError
+from charm_state import Arch, ProxyConfig, SSHDebugConnection
+from errors import OpenStackUnauthorizedError, RunnerBinaryError
+from github_client import GithubClient
+from github_type import RunnerApplication
+from runner_type import GithubPath
+from utilities import execute_command, retry
 
 logger = logging.getLogger(__name__)
 
-CLOUDS_YAML_PATH = Path(Path.home() / ".config/openstack/clouds.yaml")
-
-
-def _validate_cloud_config(cloud_config: dict) -> None:
-    """Validate the format of the cloud configuration.
-
-    Args:
-        cloud_config: The configuration in clouds.yaml format to validate.
-
-    Raises:
-        InvalidConfigError: if the format of the config is invalid.
-    """
-    # dict of format: {clouds: <cloud-name>: <cloud-config>}
-    try:
-        clouds = list(cloud_config["clouds"].keys())
-    except KeyError as exc:
-        raise OpenStackInvalidConfigError("Invalid clouds.yaml.") from exc
-
-    if not clouds:
-        raise OpenStackInvalidConfigError("No clouds defined in clouds.yaml.")
-
-
-def _write_config_to_disk(cloud_config: dict) -> None:
-    """Write the cloud configuration to disk.
+IMAGE_PATH_TMPL = "jammy-server-cloudimg-{architecture}-compressed.img"
+IMAGE_NAME = "jammy"
+BUILD_OPENSTACK_IMAGE_SCRIPT_FILENAME = "scripts/build-openstack-image.sh"
 
-    Args:
-        cloud_config: The configuration in clouds.yaml format to write to disk.
-    """
-    CLOUDS_YAML_PATH.parent.mkdir(parents=True, exist_ok=True)
-    CLOUDS_YAML_PATH.write_text(encoding="utf-8", data=yaml.dump(cloud_config))
 
-
-def _create_connection(cloud_config: dict) -> openstack.connection.Connection:
+def _create_connection(cloud_config: dict[str, dict]) -> openstack.connection.Connection:
```

Comment:
Perhaps this can be a typed dict with the `clouds` key and the value of that just an arbitrary dictionary?

Reply:
Added CloudConfig type, thank you!